tests/unit/suites/libraries/cms/router/JRouterSiteTest.php

```diff
@@ -417,211 +598,515 @@ public function testParseRawRoute($url, $menubool, $expectedVars, $expectedGloba
 
 		if (isset($appConfig['languagefilter']))
 		{
-			$app->expects($this->any())->method('getLanguageFilter')->will($this->returnValue(true));
+			$app->expects($this->any())
+				->method('getLanguageFilter')
+				->will($this->returnValue(true));
 			unset($appConfig['languagefilter']);
 		}
-		$app->expects($this->any())->method('get')->will($this->returnValueMap($appConfig));
-		$this->object->setApp($app);
+		$app->expects($this->any())
+			->method('get')
+			->will($this->returnValueMap($appConfig));
 
-		if ($menubool)
+		if ($mode == JROUTER_MODE_SEF)
 		{
-			$menu = TestMockMenu::create($this, false, $activeMenu);
-			$menu->expects($this->any())->method('getDefault')->will($this->returnValue(null));
-			$this->object->setMenu($menu);
+			$menu = TestMockMenu::create($this, false, $setActive);
+			$menu
+				->expects($this->any())
+				->method('getDefault')
+				->will($this->returnValue(null));
 		}
 		else
 		{
-			$menu = TestMockMenu::create($this, true, $activeMenu);
-			$this->object->setMenu($menu);
+			$menu = TestMockMenu::create($this, true, $setActive);
 		}
 
-		// The method should return an array of variables
-		$vars = $this->object->runParseSefRoute($uri);
-		$this->assertEquals($expected, $vars);
+		$object = new JRouterSite(
+			array(),
+			$app,
+			$menu
+		);
+
+		$parseSefRouteMethod = new ReflectionMethod('JRouterSite', 'parseSefRoute');
+		$parseSefRouteMethod->setAccessible(true);
 
-		$this->assertEquals($expectedGlobals, $this->object->getVars(), 'global vars');
+		$vars = $parseSefRouteMethod->invoke($object, new JUri($url));
+
+		$this->assertEquals(
+			$expectedParseVars,
+			$vars,
+			"JRouterSite::parseSefRoute() did not return the expected values."
+		);
+		$this->assertEquals(
+			$expectedObjectVars,
+			$object->getVars(),
+			"JRouterSite did not have the expected values internally."
+		);
 	}
 
 	/**
 	 * Tests the buildRawRoute() method
 	 *
 	 * @return  void
-	 *
+	 * @testdox JRouterSite::buildRawRoute() does not change a URL without an option
 	 * @since   3.4
 	 */
 	public function testBuildRawRoute()
 	{
 		$uri = new JUri('index.php');
 
-		// Test if a URL without an option is returned identical
-		$this->object->runBuildRawRoute($uri);
-		$this->assertEquals('index.php', $uri->toString());
-
-		// Test if a component routers preprocess method is executed
-		$uri->setVar('option', 'com_test');
-		$this->object->runBuildRawRoute($uri);
-		$this->assertEquals('index.php?option=com_test&testvar=testvalue', $uri->toString());
+		$object = new JRouterSite(
+			array(),
+			$this->getMockCmsApp(),
+			TestMockMenu::create($this)
+		);
 
-		// Test if a broken option is properly sanitised to get the right router
-		$uri->setVar('option', 'com_ te?st');
-		$uri->delVar('testvar');
-		$this->object->runBuildRawRoute($uri);
-		$this->assertEquals('index.php?option=com_ te?st&testvar=testvalue', $uri->toString());
+		$buildRawRouteMethod = new ReflectionMethod('JRouterSite', 'buildRawRoute');
```

Comment:
Does not work because of the call-by-reference problem. That's why `$method->invokeArgs($object, array(&$uri))` has to be used instead of `$method->invoke($object, &$uri)`.